### ISSUE TITLE
NS3002 false positive when nested returns

### ISSUE THREAD

**Issue by andrewdmoreno** [nsubstitute/NSubstitute.Analyzers#150]
#118 was closed so did not want to add onto it, however, I am running into this issue still. 

```` c#
var provider = Substitute.For<IOptionProvider>();

provider
    .GetOptions(Arg.Any<EnrollmentDetail>())
    .Returns(_ =>
        {
            //  NS3002 Can not find an argument of type ... to this call
            var enrollment = _.Arg<EnrollmentDetail>();
            return new List<OptionSummary>();
       });
````

IOptionProvider:
```` c# 
List<OptionSummary> GetOptions(EnrollmentDetail enrollment);
````

I tried with versions `1.0.11`, `1.0.13`, and `1.0.14`. Am I doing something wrong possibly?

**Comment by tpodolak** [comment]
Hi @andrewdmoreno 
I am not able to repro this. Here is my code based on your snippet
```c#
using System.Collections.Generic;
using NSubstitute;

namespace TestProject6
{
    public class FooTests
    {
        public void Test()
        {
            var provider = Substitute.For<IOptionProvider>();

            provider
                .GetOptions(Arg.Any<EnrollmentDetail>())
                .Returns(_ =>
                {
                    var enrollment = _.Arg<EnrollmentDetail>();
                    return new List<OptionSummary>();
                });
        }
    }

    public interface IOptionProvider
    {
        List<OptionSummary> GetOptions(EnrollmentDetail enrollment);
    }

    public class OptionSummary
    {
    }

    public class EnrollmentDetail
    {
    }
}
```
no warnings were detected with 
````
<PackageReference Include="NSubstitute" Version="4.2.1" />
<PackageReference Include="NSubstitute.Analyzers.CSharp" Version="1.0.14" />
````
Can you provide small solution with repro?

**Comment by andrewdmoreno** [comment]
Thanks @tpodolak. I will double check both package versions and see if I can repro it in a smaller solution.

**Comment by andrewdmoreno** [comment]
@tpodolak At first I could not repro in a fresh solution so I started looking at what else was unique about the setup and was finally able to repro. I reduced it down to the minimum needed to repro. 

```` c#
    public class FooTests
    {
        public void Test()
        {
            var providerFactory = Substitute.For<IProviderFactory>();
            providerFactory.GetProvider().Returns(info =>
            {
                var provider = Substitute.For<IOptionProvider>();
                provider
                    .GetOptions(Arg.Any<string>())
                    .Returns(_ =>
                    {
                        // Warning	NS3002	Can not find an argument of type string to this call.
                        var input = _.Arg<string>();
                        return new List<string> {input};
                    });
                return provider;
            });
        }
    }

    public interface IProviderFactory
    {
        IOptionProvider GetProvider();
    }

    public interface IOptionProvider
    {
        List<string> GetOptions(string input);
    }
````

**Comment by dtchepak** [comment]
@andrewdmoreno Thanks for taking the time to refine the reproduction case!

@tpodolak I'm guessing the issue is the nested `Returns` here?

**Comment by andrewdmoreno** [comment]
Now that we know that the nested `Returns` caused this false positive I have mitigated the issue on my end by moving the provider setup outside and just returning the provider instance like so:

```` c#
    public class FooTests
    {
        public void Test()
        {
            var provider = Substitute.For<IOptionProvider>();
            provider
                .GetOptions(Arg.Any<string>())
                .Returns(_ =>
                {
                    var input = _.Arg<string>();
                    return new List<string> {input};
                });

            var providerFactory = Substitute.For<IProviderFactory>();
            providerFactory.GetProvider().Returns(provider);
        }
    }

    public interface IProviderFactory
    {
        IOptionProvider GetProvider();
    }

    public interface IOptionProvider
    {
        List<string> GetOptions(string input);
    }
````

However, I imagine there could be valid use cases for having nested `Returns`

**Comment by tpodolak** [comment]
>@tpodolak I'm guessing the issue is the nested Returns here?
Yes, there is a missing check if given Arg<T> is belongs to given callInfo parameter.
